// @department-of-veterans-affairs/formation-react/Column.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _kebabCase2 = _interopRequireDefault(require("lodash/kebabCase"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var isPanelWhite = function isPanelWhite(mobileMediaQuery, panelWhite) {
  if (mobileMediaQuery.matches) {
    return '';
  }

  return panelWhite ? ' panel-white' : '';
};

var Column = function Column(props) {
  var mobileMediaQuery = props.mobileMediaQuery,
      data = props.data,
      keyName = props.keyName,
      panelWhite = props.panelWhite,
      columnThreeLinkClicked = props.columnThreeLinkClicked,
      linkClicked = props.linkClicked,
      hidden = props.hidden;

  if (keyName === 'columnThree') {
    return _react["default"].createElement("div", {
      "aria-hidden": hidden ? 'true' : 'false',
      className: "vetnav-panel vetnav-panel--submenu ".concat((0, _kebabCase2["default"])(keyName)).concat(isPanelWhite(mobileMediaQuery, panelWhite)),
      "aria-label": keyName
    }, _react["default"].createElement("div", {
      className: "".concat(panelWhite ? 'mm-marketing-container mm-marketing-gray' : 'mm-marketing-container')
    }, _react["default"].createElement("img", {
      src: data.img.src,
      alt: data.img.alt
    }), _react["default"].createElement("div", {
      className: "mm-marketing-text"
    }, _react["default"].createElement("a", {
      className: "mm-links",
      href: data.link.href,
      onClick: columnThreeLinkClicked.bind(null, data.link),
      target: data.link.target || '_self'
    }, data.link.text), _react["default"].createElement("p", null, data.description))));
  }

  return _react["default"].createElement("div", {
    "aria-hidden": hidden ? 'true' : 'false',
    className: "vetnav-panel vetnav-panel--submenu ".concat((0, _kebabCase2["default"])(keyName)).concat(isPanelWhite(mobileMediaQuery, panelWhite))
  }, data.title ? _react["default"].createElement("h3", {
    id: "vetnav-".concat((0, _kebabCase2["default"])(keyName), "-header")
  }, data.title) : _react["default"].createElement("span", null, "\xA0"), _react["default"].createElement("ul", {
    id: "vetnav-".concat((0, _kebabCase2["default"])(keyName), "-col"),
    "aria-labelledby": "vetnav-".concat((0, _kebabCase2["default"])(keyName), "-header")
  }, _react["default"].createElement("li", {
    className: "panel-top-link"
  }, props.children), data.links.map(function (link, i) {
    return _react["default"].createElement("li", {
      className: "mm-link-container",
      key: "".concat(link.href, "-").concat(i)
    }, _react["default"].createElement("a", {
      className: "mm-links",
      href: link.href,
      onClick: linkClicked.bind(null, link),
      target: link.target || '_self'
    }, link.text));
  })));
};

var _default = Column;
exports["default"] = _default;
