// @department-of-veterans-affairs/formation-react/ErrorableFileInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _uniqueId2 = _interopRequireDefault(require("lodash/uniqueId"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var ErrorableFileInput = /*#__PURE__*/function (_React$Component) {
  _inherits(ErrorableFileInput, _React$Component);

  function ErrorableFileInput() {
    var _this;

    _classCallCheck(this, ErrorableFileInput);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(ErrorableFileInput).call(this));
    _this.handleChange = _this.handleChange.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(ErrorableFileInput, [{
    key: "UNSAFE_componentWillMount",
    value: function UNSAFE_componentWillMount() {
      this.inputId = (0, _uniqueId2["default"])('errorable-file-input-');
    }
  }, {
    key: "handleChange",
    value: function handleChange(domEvent) {
      this.props.onChange(domEvent.target.files); // clear the original input, otherwise events will be triggered
      // with empty file arrays and sometimes uploading a file twice will
      // not work

      domEvent.target.value = null; // eslint-disable-line no-param-reassign
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var errorSpan = '';
      var errorSpanId = undefined;
      var inputErrorClass = undefined;
      var labelErrorClass = undefined;

      if (this.props.errorMessage) {
        errorSpanId = "".concat(this.inputId, "-error-message");
        errorSpan = _react["default"].createElement("span", {
          className: "usa-input-error-message ".concat(this.props.additionalErrorClass),
          role: "alert",
          id: errorSpanId
        }, _react["default"].createElement("span", {
          className: "sr-only"
        }, "Error"), " ", this.props.errorMessage);
        inputErrorClass = 'usa-input-error';
        labelErrorClass = 'usa-input-error-label';
      } // Calculate required


      var requiredSpan = undefined;

      if (this.props.required) {
        requiredSpan = _react["default"].createElement("span", {
          className: "form-required-span"
        }, "(*Required)");
      }

      return _react["default"].createElement("div", {
        className: this.props.additionalClass
      }, _react["default"].createElement("div", {
        className: inputErrorClass,
        role: "alert"
      }, _react["default"].createElement("label", {
        className: labelErrorClass,
        htmlFor: this.inputId
      }, this.props.label, requiredSpan), errorSpan, _react["default"].createElement("label", {
        role: "button",
        tabIndex: "0",
        htmlFor: this.inputId,
        onKeyPress: function onKeyPress(e) {
          if (e.key === 'Enter' || e.key === ' ') {
            document.getElementById(_this2.inputId).click();
          }
        },
        "aria-describedby": this.props['aria-describedby'],
        className: this.props.triggerClass || 'usa-button usa-button-secondary'
      }, this.props.buttonText), _react["default"].createElement("input", {
        multiple: this.props.multiple,
        style: {
          display: 'none'
        },
        type: "file",
        accept: this.props.accept,
        id: this.inputId,
        name: this.props.name,
        onChange: this.handleChange
      })));
    }
  }]);

  return ErrorableFileInput;
}(_react["default"].Component);

ErrorableFileInput.defaultProps = {
  buttonText: 'Add Files',
  mimeTypes: '',
  multiple: false
};
var _default = ErrorableFileInput;
exports["default"] = _default;
