// @department-of-veterans-affairs/formation-react/MaintenanceBanner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.MaintenanceBanner = exports.MAINTENANCE_BANNER = void 0;

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _moment = _interopRequireDefault(require("moment"));

var _AlertBox = _interopRequireDefault(require("./AlertBox"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var MAINTENANCE_BANNER = 'MAINTENANCE_BANNER'; // @WARNING: This is currently only used once in vets-website.

/**
 * Display a maintenance banner for a given time window.
 */

exports.MAINTENANCE_BANNER = MAINTENANCE_BANNER;

var MaintenanceBanner = /*#__PURE__*/function (_Component) {
  _inherits(MaintenanceBanner, _Component);

  function MaintenanceBanner(props) {
    var _this;

    _classCallCheck(this, MaintenanceBanner);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(MaintenanceBanner).call(this, props));

    _defineProperty(_assertThisInitialized(_this), "derivePostContent", function () {
      var _this$props = _this.props,
          startsAt = _this$props.startsAt,
          expiresAt = _this$props.expiresAt;
      var startsAtET = startsAt.clone().subtract(4, 'hours');
      var expiresAtET = expiresAt.clone().subtract(4, 'hours');

      if (startsAt.isSame(expiresAt, 'day')) {
        return _react["default"].createElement(_react["default"].Fragment, null, _react["default"].createElement("p", null, _react["default"].createElement("strong", null, "Date:"), " ", startsAtET.format('dddd MMMM D, YYYY')), _react["default"].createElement("p", null, _react["default"].createElement("strong", null, "Start/End time:"), " ", startsAtET.format('h:mm a'), " to", ' ', expiresAtET.format('h:mm a'), " ET"));
      }

      return _react["default"].createElement(_react["default"].Fragment, null, _react["default"].createElement("p", null, _react["default"].createElement("strong", null, "Start:"), ' ', startsAtET.format('dddd MMMM D, YYYY, [at] h:mm a'), " ET"), _react["default"].createElement("p", null, _react["default"].createElement("strong", null, "End:"), ' ', expiresAtET.format('dddd MMMM D, YYYY, [at] h:mm a'), " ET"));
    });

    _defineProperty(_assertThisInitialized(_this), "onCloseAlert", function () {
      if (_this.props.localStorage) {
        _this.props.localStorage.setItem(MAINTENANCE_BANNER, _this.props.id);
      }

      _this.setState({
        dismissed: true
      });
    });

    _this.state = {
      dismissed: props.localStorage && props.localStorage.getItem(MAINTENANCE_BANNER) === _this.props.id
    };
    return _this;
  }

  _createClass(MaintenanceBanner, [{
    key: "render",
    value: function render() {
      var derivePostContent = this.derivePostContent,
          onCloseAlert = this.onCloseAlert;
      var dismissed = this.state.dismissed;
      var _this$props2 = this.props,
          content = _this$props2.content,
          expiresAt = _this$props2.expiresAt,
          id = _this$props2.id,
          startsAt = _this$props2.startsAt,
          title = _this$props2.title,
          warnContent = _this$props2.warnContent,
          warnStartsAt = _this$props2.warnStartsAt,
          warnTitle = _this$props2.warnTitle; // Derive dates.

      var now = (0, _moment["default"])();
      var postContent = derivePostContent(); // Escape early if the banner is dismissed.

      if (dismissed) {
        return null;
      } // Escape early if it's before when it should show.


      if (now.isBefore(warnStartsAt)) {
        return null;
      } // Escape early if it's after when it should show.


      if (now.isAfter(expiresAt)) {
        return null;
      } // Show pre-downtime.


      if (now.isBefore(startsAt)) {
        return _react["default"].createElement("div", {
          className: "usa-alert-full-width vads-u-border-top--5px medium-screen:vads-u-border-top--10px vads-u-border-color--warning-message maintenance-banner",
          "data-e2e-id": "maintenance-banner-pre-downtime"
        }, _react["default"].createElement(_AlertBox["default"], {
          content: _react["default"].createElement(_react["default"].Fragment, null, _react["default"].createElement("p", null, warnContent), postContent),
          headline: warnTitle,
          onCloseAlert: onCloseAlert,
          status: "warning"
        }));
      } // Show downtime.


      return _react["default"].createElement("div", {
        className: "usa-alert-full-width vads-u-border-top--5px medium-screen:vads-u-border-top--10px vads-u-border-color--secondary maintenance-banner",
        "data-e2e-id": "maintenance-banner-downtime"
      }, _react["default"].createElement(_AlertBox["default"], {
        content: _react["default"].createElement(_react["default"].Fragment, null, _react["default"].createElement("p", null, content), postContent),
        headline: title,
        onCloseAlert: onCloseAlert,
        status: "error"
      }));
    }
  }]);

  return MaintenanceBanner;
}(_react.Component);

exports.MaintenanceBanner = MaintenanceBanner;
var _default = MaintenanceBanner;
exports["default"] = _default;
