// @department-of-veterans-affairs/formation-react/OMBInfo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _Modal = _interopRequireDefault(require("./Modal"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var OMBInfo = /*#__PURE__*/function (_React$Component) {
  _inherits(OMBInfo, _React$Component);

  function OMBInfo(props) {
    var _this;

    _classCallCheck(this, OMBInfo);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(OMBInfo).call(this, props));

    _defineProperty(_assertThisInitialized(_this), "openModal", function () {
      _this.setState({
        modalOpen: true
      });
    });

    _defineProperty(_assertThisInitialized(_this), "closeModal", function () {
      _this.setState({
        modalOpen: false
      });
    });

    _defineProperty(_assertThisInitialized(_this), "modalContents", function (minutes) {
      return _react["default"].createElement("div", null, _react["default"].createElement("h3", null, "Privacy Act Statement"), minutes && _react["default"].createElement("p", null, _react["default"].createElement("strong", null, "Respondent Burden:"), " We need this information to determine your eligibility for education benefits (38 U.S.C. 3471). Title 38, United States Code, allows us to ask for this information. We estimate that you will need an average of ", minutes, " minutes to review the instructions, find the information, and complete this form. The VA cannot conduct or sponsor a collection of information unless a valid OMB (Office of Management and Budget) control number is displayed. You are not required to respond to a collection of information if this number is not displayed. Valid OMB control numbers can be located on the OMB Internet Page at www.reginfo.gov/public/do/PRAMain. If desired, you can call", ' ', _react["default"].createElement("a", {
        href: "+18008271000"
      }, "1-800-827-1000"), " to get information on where to send comments or suggestions about this form."), _react["default"].createElement("p", null, _react["default"].createElement("strong", null, "Privacy Act Notice:"), " The VA will not disclose information collected on this form to any source other than what has been authorized under the Privacy Act of 1974 or title 38, Code of Federal Regulations, section 1.576 for routine uses (e.g., the VA sends educational forms or letters with a veteran\u2019s identifying information to the Veteran\u2019s school or training establishment to (1) assist the veteran in the completion of claims forms or (2) for the VA to obtain further information as may be necessary from the school for VA to properly process the Veteran\u2019s education claim or to monitor his or her progress during training) as identified in the VA system of records, 58VA21/22/28, Compensation, Pension, Education, and Vocational Rehabilitation and Employment Records - VA, and published in the Federal Register. Your obligation to respond is required to obtain or retain education benefits. Giving us your SSN account information is voluntary. Refusal to provide your SSN by itself will not result in the denial of benefits. The VA will not deny an individual benefits for refusing to provide his or her SSN unless the disclosure of the SSN is required by a Federal Statute of law enacted before January 1, 1975, and still in effect. The requested information is considered relevant and necessary to determine the maximum benefits under the law. While you do not have to respond, VA cannot process your claim for education assistance unless the information is furnished as required by existing law (38 U.S.C. 3471). The responses you submit are considered confidential (38 U.S.C. 5701). Any information provided by applicants, recipients, and others may be subject to verification through computer matching programs with other agencies."));
    });

    _this.state = {
      modalOpen: false
    };
    return _this;
  }

  _createClass(OMBInfo, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          resBurden = _this$props.resBurden,
          ombNumber = _this$props.ombNumber,
          expDate = _this$props.expDate;
      return _react["default"].createElement("div", {
        className: "omb-info"
      }, resBurden && _react["default"].createElement("div", null, "Respondent burden: ", _react["default"].createElement("strong", null, resBurden, " minutes")), ombNumber && _react["default"].createElement("div", null, "OMB Control #: ", _react["default"].createElement("strong", null, ombNumber)), _react["default"].createElement("div", null, "Expiration date: ", _react["default"].createElement("strong", null, expDate)), _react["default"].createElement("div", null, _react["default"].createElement("button", {
        className: "va-button-link",
        onClick: this.openModal
      }, "Privacy Act Statement")), _react["default"].createElement(_Modal["default"], {
        cssClass: "va-modal-large",
        contents: this.props.children ? this.props.children : this.modalContents(resBurden),
        id: "omb-modal",
        visible: this.state.modalOpen,
        onClose: this.closeModal
      }));
    }
  }]);

  return OMBInfo;
}(_react["default"].Component);

var _default = OMBInfo;
exports["default"] = _default;
