// @department-of-veterans-affairs/formation-react/options-for-select.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.yesNo = exports.childRelationships = exports.days = exports.thirtyOneDays = exports.thirtyDays = exports.twentyNineDays = exports.months = exports.genders = exports.suffixes = exports.dischargeTypes = exports.branchesServed = exports.maritalStatuses = void 0;
var maritalStatuses = ['Married', 'Never Married', 'Separated', 'Widowed', 'Divorced'];
exports.maritalStatuses = maritalStatuses;
var branchesServed = [{
  value: 'air force',
  label: 'Air Force'
}, {
  value: 'army',
  label: 'Army'
}, {
  value: 'coast guard',
  label: 'Coast Guard'
}, {
  value: 'marine corps',
  label: 'Marine Corps'
}, {
  value: 'merchant seaman',
  label: 'Merchant Seaman'
}, {
  value: 'navy',
  label: 'Navy'
}, {
  value: 'noaa',
  label: 'Noaa'
}, {
  value: 'usphs',
  label: 'USPHS'
}, {
  value: 'f.commonwealth',
  label: 'Filipino Commonwealth Army'
}, {
  value: 'f.guerilla',
  label: 'Filipino Guerilla Forces'
}, {
  value: 'f.scouts new',
  label: 'Filipino New Scout'
}, {
  value: 'f.scouts old',
  label: 'Filipino Old Scout'
}, {
  value: 'other',
  label: 'Other'
}];
exports.branchesServed = branchesServed;
var dischargeTypes = [{
  value: 'honorable',
  label: 'Honorable'
}, {
  value: 'general',
  label: 'General'
}, {
  value: 'other',
  label: 'Other Than Honorable'
}, {
  value: 'bad-conduct',
  label: 'Bad Conduct'
}, {
  value: 'dishonorable',
  label: 'Dishonorable'
}, {
  value: 'undesirable',
  label: 'Undesirable'
}];
exports.dischargeTypes = dischargeTypes;
var suffixes = ['Jr.', 'Sr.', 'II', 'III', 'IV'];
exports.suffixes = suffixes;
var genders = [{
  label: 'Female',
  value: 'F'
}, {
  label: 'Male',
  value: 'M'
}];
exports.genders = genders;
var months = [{
  label: 'Jan',
  value: 1
}, {
  label: 'Feb',
  value: 2
}, {
  label: 'Mar',
  value: 3
}, {
  label: 'Apr',
  value: 4
}, {
  label: 'May',
  value: 5
}, {
  label: 'Jun',
  value: 6
}, {
  label: 'Jul',
  value: 7
}, {
  label: 'Aug',
  value: 8
}, {
  label: 'Sep',
  value: 9
}, {
  label: 'Oct',
  value: 10
}, {
  label: 'Nov',
  value: 11
}, {
  label: 'Dec',
  value: 12
}];
exports.months = months;
var twentyNineDays = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29'];
exports.twentyNineDays = twentyNineDays;
var thirtyDays = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30'];
exports.thirtyDays = thirtyDays;
var thirtyOneDays = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31'];
exports.thirtyOneDays = thirtyOneDays;
var days = {
  1: thirtyOneDays,
  2: twentyNineDays,
  3: thirtyOneDays,
  4: thirtyDays,
  5: thirtyOneDays,
  6: thirtyDays,
  7: thirtyOneDays,
  8: thirtyOneDays,
  9: thirtyDays,
  10: thirtyOneDays,
  11: thirtyDays,
  12: thirtyOneDays
};
exports.days = days;
var childRelationships = ['Daughter', 'Son', 'Stepson', 'Stepdaughter'];
exports.childRelationships = childRelationships;
var yesNo = [{
  label: 'Yes',
  value: 'Y'
}, {
  label: 'No',
  value: 'N'
}];
exports.yesNo = yesNo;
