// @department-of-veterans-affairs/formation-react/PrivacyAgreement.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = PrivacyAgreement;

var _react = _interopRequireDefault(require("react"));

var _ErrorableCheckbox = _interopRequireDefault(require("./ErrorableCheckbox"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function PrivacyAgreement(_ref) {
  var onChange = _ref.onChange,
      checked = _ref.checked,
      showError = _ref.showError;
  return _react["default"].createElement("div", null, _react["default"].createElement(_ErrorableCheckbox["default"], {
    required: true,
    checked: checked,
    onValueChange: onChange,
    name: "privacyAgreement",
    errorMessage: showError && !checked ? 'You must accept the privacy policy before continuing' : undefined,
    label: _react["default"].createElement("span", null, "I have read and accept the", ' ', _react["default"].createElement("a", {
      target: "_blank",
      href: "/privacy/"
    }, "privacy policy"))
  }));
}
