// @department-of-veterans-affairs/formation-react/PromoBanner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PROMO_BANNER_TYPES = exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var PROMO_BANNER_TYPES = {
  announcement: 'announcement',
  news: 'news',
  emailSignup: 'email-signup'
};
exports.PROMO_BANNER_TYPES = PROMO_BANNER_TYPES;
var PROMO_BANNER_ICONS = new Map([[PROMO_BANNER_TYPES.announcement, 'fa-bullhorn'], [PROMO_BANNER_TYPES.news, 'fa-newspaper'], [PROMO_BANNER_TYPES.emailSignup, 'fa-envelope']]);

function PromoBanner(_ref) {
  var type = _ref.type,
      onClose = _ref.onClose,
      render = _ref.render,
      href = _ref.href,
      target = _ref.target,
      text = _ref.text;
  var iconClasses = (0, _classnames["default"])('fas', 'fa-stack-1x', PROMO_BANNER_ICONS.get(type));
  return _react["default"].createElement("div", {
    className: "vads-c-promo-banner"
  }, _react["default"].createElement("div", {
    className: "vads-c-promo-banner__body"
  }, _react["default"].createElement("div", {
    className: "vads-c-promo-banner__icon"
  }, _react["default"].createElement("span", {
    className: "fa-stack fa-lg"
  }, _react["default"].createElement("i", {
    className: "vads-u-color--white fa fa-circle fa-stack-2x"
  }), _react["default"].createElement("i", {
    className: iconClasses
  }))), _react["default"].createElement("div", {
    className: "vads-c-promo-banner__content"
  }, render ? render() : _react["default"].createElement("a", {
    className: "vads-c-promo-banner__content-link",
    href: href,
    target: target,
    onClick: onClose
  }, text, " ", _react["default"].createElement("i", {
    className: "fas fa-angle-right"
  }))), _react["default"].createElement("div", {
    className: "vads-c-promo-banner__close"
  }, _react["default"].createElement("button", {
    type: "button",
    "aria-label": "Dismiss this announcement",
    onClick: onClose,
    className: "va-button-link vads-u-margin-top--1"
  }, _react["default"].createElement("i", {
    className: "fas fa-times-circle vads-u-font-size--lg"
  })))));
}

var _default = PromoBanner;
exports["default"] = _default;
