// @department-of-veterans-affairs/formation-react/SearchMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _IconSearch = _interopRequireDefault(require("./IconSearch"));

var _DropDownPanel = _interopRequireDefault(require("./DropDownPanel"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var SearchMenu = /*#__PURE__*/function (_React$Component) {
  _inherits(SearchMenu, _React$Component);

  function SearchMenu(props) {
    var _this;

    _classCallCheck(this, SearchMenu);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(SearchMenu).call(this, props));

    _defineProperty(_assertThisInitialized(_this), "toggleSearchForm", function () {
      _this.props.clickHandler();
    });

    _defineProperty(_assertThisInitialized(_this), "makeForm", function () {
      return _react["default"].createElement("form", {
        acceptCharset: "UTF-8",
        action: "https://search.vets.gov/search",
        id: "search",
        method: "get"
      }, _react["default"].createElement("div", {
        className: "csp-inline-patch-header"
      }, _react["default"].createElement("input", {
        name: "utf8",
        type: "hidden",
        value: "\u2713"
      })), _react["default"].createElement("input", {
        id: "affiliate",
        name: "affiliate",
        type: "hidden",
        value: "vets.gov_search"
      }), _react["default"].createElement("label", {
        htmlFor: "query",
        className: "usa-sr-only"
      }, "Search:"), _react["default"].createElement("div", {
        className: "va-flex"
      }, _react["default"].createElement("input", {
        autoComplete: "off",
        ref: _this.searchFieldRef,
        className: "usagov-search-autocomplete",
        id: "query",
        name: "query",
        type: "text"
      }), _react["default"].createElement("button", {
        type: "submit"
      }, _react["default"].createElement(_IconSearch["default"], {
        color: "#fff",
        role: "presentation"
      }), _react["default"].createElement("span", {
        className: "usa-sr-only"
      }, "Search"))));
    });

    _this.searchFieldRef = _react["default"].createRef();
    return _this;
  }

  _createClass(SearchMenu, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      this.searchFieldRef.current.focus();
    }
  }, {
    key: "render",
    value: function render() {
      var buttonClasses = (0, _classnames["default"])(this.props.cssClass, 'va-btn-withicon', '');

      var icon = _react["default"].createElement(_IconSearch["default"], {
        color: "#fff",
        role: "presentation"
      });

      return _react["default"].createElement(_DropDownPanel["default"], {
        buttonText: "Search",
        clickHandler: this.props.clickHandler,
        cssClass: buttonClasses,
        id: "searchmenu",
        icon: icon,
        isOpen: this.props.isOpen
      }, this.makeForm());
    }
  }]);

  return SearchMenu;
}(_react["default"].Component);

var _default = SearchMenu;
exports["default"] = _default;
