// @department-of-veterans-affairs/formation-react/SubMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _kebabCase2 = _interopRequireDefault(require("lodash/kebabCase"));

var _react = _interopRequireDefault(require("react"));

var _reactFocusLock = _interopRequireDefault(require("react-focus-lock"));

var _Column = _interopRequireDefault(require("./Column"));

var _arrowRightBlue = _interopRequireDefault(require("./arrow-right-blue"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

var getColumns = function getColumns(mobileMediaQuery, columns) {
  if (mobileMediaQuery.matches) {
    return {
      columnOne: {
        title: columns.columnOne.title,
        links: [].concat(_toConsumableArray(columns.columnOne.links), _toConsumableArray(columns.columnTwo.links))
      }
    };
  }

  return columns;
};

var SubMenu = function SubMenu(_ref) {
  var mobileMediaQuery = _ref.mobileMediaQuery,
      smallDesktopMediaQuery = _ref.smallDesktopMediaQuery,
      data = _ref.data,
      id = _ref.id,
      show = _ref.show,
      navTitle = _ref.navTitle,
      handleBackToMenu = _ref.handleBackToMenu,
      linkClicked = _ref.linkClicked,
      columnThreeLinkClicked = _ref.columnThreeLinkClicked;

  var seeAllLink = data.seeAllLink,
      columns = _objectWithoutProperties(data, ["seeAllLink"]);

  if (show) {
    var filteredColumns = getColumns(mobileMediaQuery, columns);
    return _react["default"].createElement("div", {
      className: mobileMediaQuery.matches ? 'mm-link-container-small' : '',
      id: id,
      role: "group"
    }, _react["default"].createElement(_reactFocusLock["default"], {
      disabled: !mobileMediaQuery.matches
    }, _react["default"].createElement("div", null, _react["default"].createElement("button", {
      className: "back-button",
      "aria-controls": "vetnav-".concat((0, _kebabCase2["default"])(navTitle)),
      onClick: function onClick() {
        return handleBackToMenu();
      }
    }, "Back to Menu")), seeAllLink && _react["default"].createElement("div", {
      className: "panel-bottom-link"
    }, _react["default"].createElement("a", {
      href: seeAllLink.href,
      onClick: linkClicked.bind(null, seeAllLink)
    }, seeAllLink.text, _react["default"].createElement(_arrowRightBlue["default"], null))), Object.keys(filteredColumns).map(function (keyName) {
      return _react["default"].createElement(_Column["default"], {
        key: keyName,
        data: filteredColumns[keyName],
        keyName: keyName,
        navTitle: navTitle,
        panelWhite: Object.prototype.hasOwnProperty.call(filteredColumns, 'mainColumn'),
        linkClicked: linkClicked,
        mobileMediaQuery: mobileMediaQuery,
        hidden: keyName === 'columnThree' && smallDesktopMediaQuery.matches,
        columnThreeLinkClicked: columnThreeLinkClicked
      });
    })));
  }

  return _react["default"].createElement("div", null);
};

var _default = SubMenu;
exports["default"] = _default;
